### PR TITLE
Add camera sensor simulation 

### PR DESCRIPTION
**Pull request by juan-g-bonilla**
# PR — `feat(sensors): backend-agnostic mounted cameras + camera-egress plugins`

## Summary

Adds a **backend-agnostic mounted-camera system** and the **camera-egress plugins** that ship rendered
frames off the sim. Declare any number of cameras in one top-level `sensor:` config; mount them on a
robot link, a scene actor, or the world; render RGB and/or depth on all four backends (MuJoCo Classic
CPU, MuJoCo Warp GPU, IsaacGym, IsaacSim); read frames through one unified `get_camera_data` API with a
shared cross-device cache; feed them to visual-policy **observation terms**; and stream them out via
**plugins** (`plugin.<key>:ros2-image` / `viz` / …) — ROS2 `Image`/`CompressedImage` publish, a live cv2
window, or an mp4. Built on the plugin family (base branch): camera egress is a plugin like any other, so
it works in both `run_sim` and training. This is the sensor/vision layer of the stack.

### Newly supported (opt-in; no-op default — no existing task/env affected)

- **`sensor:` camera config family, selectable as `--sensor.<name>:<variant>`.** New
  `CameraSensorConfig` (`config_types/sensor.py:...`) — backend-agnostic core fields
  (width/height/vertical_fov/near/far/data_types/update_decimation) + a `SensorMountConfig`
  (`robot_link` / `actor` / `world`) + per-backend sub-configs — resolved from `CAMERA_REGISTRY`
  (`config_values/sensor.py:27`, group `holosoma.config.sensor`). Surfaced as a dynamic-dict field:
  `sensor` on `RunSimConfig` (`run_sim.py:83`) / `ExperimentConfig` (`experiment.py:129`) /
  `EnvConfig` (`env.py:30`), threaded into `FullSimConfig.sensors` (`full_sim.py:34`, plural). Empty by
  default. G1 building-block cameras (`g1-head`, `g1-stereo-left/right`, `g1-*-wrist`, `g1-waist-*`)
  ship in `config_values/wbt/g1/sensor.py` — composed per-key on the CLI.
- **Per-backend camera rendering, unified read.** Each backend renders due cameras into a
  `SensorManager`/`CameraRuntime` registry (`simulator/shared/camera_sensor.py`) with a lazy
  cross-device cache; frames are read back through `BaseSimulator.get_camera_data(name, modality,
  device=…)` (`base_simulator.py:772`) — `uint8 [N,H,W,3]` R,G,B for rgb, `float32 [N,H,W,1]` meters
  (image-plane, `+inf` no-hit) for depth. IsaacSim uses its native tiled camera; MuJoCo-Warp uses its
  batched renderer; IsaacGym / MuJoCo-Classic render per camera. `render_sensors` runs once per control
  frame as a core `FRAME_END` hook, registered **before** any consumer plugin so buffers are fresh on
  read (`base_simulator.py:183`).
- **Camera observation terms** — `camera_rgb` / `camera_depth`
  (`managers/observation/terms/cameras.py:32,50`) read `get_camera_data` and apply an optional
  `ImageTransform` (`simulator/shared/image_transform.py`: resize → scale → layout → flatten) for
  visual policies; the observation manager handles image-shaped (non-flat) terms
  (`managers/observation/manager.py`).
- **Camera-egress plugins** — a `CameraConsumerPlugin` base
  (`simulator/plugins/camera_consumer.py:85`) self-serves each step's fresh frames (one shared
  `get_camera_data(device="cpu")` host copy per (camera, modality), sliced per wanted env) and hands them
  to concrete sinks; registers `publish` on `FRAME_END` + `stop` on `CLOSE`. Concrete plugins:
  - **`ros2-image` / `ros2-stereo` / `ros2-waist-depth` / `-color` / `-raw+color`** — publish camera
    routes as `sensor_msgs/Image` or `CompressedImage` (+ latched `CameraInfo`), one node fanning out to
    many routes, with an async drop-oldest worker per route
    (`simulator/plugins/ros2/ros2_image_plugin.py:...`, presets `config_values/plugin.py:48-143`). Depth
    routes publish raw metric (`32FC1`/`16UC1`) or colorize to RGB.
  - **`viz` / `viz-record`** — tile watched (camera, modality, env) panels into a live cv2 window and/or
    an mp4 at teardown (`simulator/plugins/viz/viz_plugin.py:...`, presets `config_values/plugin.py:145,148`).
  - ROS-free helpers (`encode`, `camera_info`, `worker`, `depth_color`, `image_grid`) live beside the
    impls; `rclpy`/`cv2` load only when a preset's `get_cls()` fires, so the config layer stays
    import-light.

### Removed / fail-loud (config surface)

- **Egress plugins validate their wanted streams at construction.** A route naming a camera/modality/env
  the sim doesn't provide raises `ValueError` at build time (`camera_consumer.py` stream validation), not
  mid-run. Per-step publish failures are isolated (logged, never propagated to the sim loop).
- **Cross-camera constraints validated** — e.g. MuJoCo-Warp render-flag agreement across the assembled
  camera dict (`config_types/sensor.py` `validate_camera_dict`), enforced at the config boundary.
- **ROS2 image routes validate format/modality** at construction (an rgb route needs an rgb format; a
  depth route takes a depth format for raw metric or an rgb format to colorize;
  `config_types/plugin.py` route validator).

### Latent / edge-case

- **`render()` is driven every step on all backends** (the run_sim/step loop calls it); headless it is a
  no-op that must tolerate a `None` viewer. Cameras render via `render_sensors`, independent of the
  interactive viewer.
- **Depth `+inf` no-hit sentinel** flows through `get_camera_data` untouched; the image-transform and
  colorize paths handle it (clip/scale), and the obs-term clip path skips image tensors so it isn't
  collapsed to a clip limit.

## Migration guide

**Users who declare no cameras see no change** — `sensor:` defaults empty and the whole path is inert.
Camera/egress authors:

**1 · Declare cameras as a dynamic-dict field**
```bash
# one head RGB camera on the robot, published over ROS2
python -m holosoma.run_sim --sensor.head_cam:g1-head plugin.pub:ros2-image

# stereo head + tile a live viz window; per-key leaf overrides
python -m holosoma.run_sim --sensor.head_cam_left:g1-stereo-left --sensor.head_cam_right:g1-stereo-right \
    --sensor.head_cam_left.width 224 plugin.viz:viz
```
The dict key is the sensor name (the `get_camera_data` key); the variant after `:` is the registry preset.
Egress-plugin routes reference cameras by that name.

**2 · Read frames in code**
```python
rgb   = sim.get_camera_data("head_cam", "rgb",   device="cpu")   # uint8  [N,H,W,3] R,G,B
depth = sim.get_camera_data("head_cam", "depth", device="cpu")   # float32 [N,H,W,1] meters, +inf no-hit
```

**3 · Visual-policy observations** — add `camera_rgb` / `camera_depth` obs terms
(`managers/observation/terms/cameras.py`), keyed to a sensor name, with an optional `ImageTransform`
(resize/scale/layout/flatten).

**4 · Write your own egress sink** — subclass `CameraConsumerPlugin`
(`simulator/plugins/camera_consumer.py`): declare `wanted_streams()`, implement `start` / `publish` /
`stop`; pair with a `PluginConfig` in `PLUGIN_REGISTRY`. Select it as `plugin.<key>:<preset>` like any
plugin. ROS2/cv2 sinks import their transport lazily.

**5 · New config fields (additive)** — `sensor: dict[str, CameraSensorConfig]` on
`RunSim`/`Experiment`/`Env` (threaded into `FullSimConfig.sensors`, plural), default `{}`; camera-egress
presets register in the existing `PLUGIN_REGISTRY` alongside `clock_publish`/`gantry_control`/`odometry`.
No existing config needs an edit.

### CODE REVIEW

File: src/holosoma/holosoma/config_types/sensor.py [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+    """(Warp, global) Apply textures; ``None`` keeps the renderer default (on). Ignored by classic."""
+
+    use_precomputed_rays: bool | None = None
+    """(Warp, global) Precompute camera rays; set ``False`` to allow per-step intrinsics
+    domain-randomization. ``None`` keeps the renderer default (True). Ignored by classic."""
+
+
+@dataclass(frozen=True, config=_FORBID_EXTRA)
+class SensorMountConfig:
+    """Where a sensor is mounted and its fixed offset (on that body, or in the per-env frame)."""
```
Comment (samuelgundry) [inline src/holosoma/holosoma/config_types/sensor.py:161]: Eventually we should look at considering at URDF/MCJF as a primary robot sim config for sensors.

File: src/holosoma/holosoma/simulator/shared/image_transform.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    # interpolate wants NCHW; round-trip the channel axis.
+    nchw = image.permute(0, 3, 1, 2).float()
+    mode = "nearest" if modality == "depth" else "bilinear"
+    kwargs = {"align_corners": False} if mode == "bilinear" else {}
+    resized = torch.nn.functional.interpolate(nchw, size=(size[0], size[1]), mode=mode, **kwargs)
+    out = resized.permute(0, 2, 3, 1)
+    return out.round().to(image.dtype) if modality == "rgb" else out
+
+
+def _scale(image: torch.Tensor, config: ImageTransformConfig, modality: str) -> torch.Tensor:
```
Comment (samuelgundry) [inline src/holosoma/holosoma/simulator/shared/image_transform.py:61]: It'd be nice to have "something" using `ImageTransformConfig`, whether as a integ-like test or config_value or sample. Not sure.

File: src/holosoma/holosoma/simulator/isaacgym/isaacgym.py [hunk 3]
```diff
@@ -498,6 +511,139 @@ def _build_each_env(self, env_id, env_ptr):
...
+        self.gym.fetch_results(self.sim, True)
+        self.gym.step_graphics(self.sim)
+        self.gym.render_all_camera_sensors(self.sim)
+        self.gym.start_access_image_tensors(self.sim)
+        try:
+            for runtime in due:
+                handles = self._camera_handles[runtime.name]  # per-env handles (parallel to self.envs)
+                if "rgb" in runtime.config.data_types:
+                    frames = []
+                    for e in range(self.num_envs):
```
Comment (samuelgundry) [inline src/holosoma/holosoma/simulator/isaacgym/isaacgym.py:637]: surprised Gym doesn't have vectorized access for sensor data. :/

File: src/holosoma/holosoma/simulator/isaacgym/isaacgym.py [hunk 4]
```diff
@@ -498,6 +511,139 @@ def _build_each_env(self, env_id, env_ptr):
...
+            return
+        # Update the graphics scene from the latest physics state, then render all camera sensors
+        # in one pass. fetch_results + step_graphics are required before render or the image is
+        # blank (same sequence the video recorder uses).
+        self.gym.fetch_results(self.sim, True)
+        self.gym.step_graphics(self.sim)
+        self.gym.render_all_camera_sensors(self.sim)
+        self.gym.start_access_image_tensors(self.sim)
+        try:
+            for runtime in due:
```
Comment (samuelgundry) [inline src/holosoma/holosoma/simulator/isaacgym/isaacgym.py:633]: nit; intuitive reading would be `for camera in ready_cameras`  or `camera_handle`

File: src/holosoma/holosoma/simulator/isaacsim/isaacsim.py [hunk 5]
```diff
@@ -513,6 +527,122 @@ def _setup_scene(self) -> None:
...
+        (``scene.update`` in ``simulate_at_each_physics_step``); this reads that output, drops
+        alpha, and writes a ``[N,H,W,3]`` uint8 frame once per control step (honoring
+        ``update_decimation``), mirroring the other backends' render -> buffer -> read flow."""
+        if self.sensor_manager is None:
+            return
+        for runtime in self.sensor_manager.collect_due():
+            out = self._tiled_cameras[runtime.name].data.output
+            if "rgb" in runtime.config.data_types:
+                rgb = out["rgb"][..., :3]  # [N,H,W,3], drop alpha
+                if rgb.dtype != torch.uint8:
```
Comment (samuelgundry) [inline src/holosoma/holosoma/simulator/isaacsim/isaacsim.py:636]: surprised you need this; IsaacSim should managed this if configured for rgb8?

File: src/holosoma/holosoma/simulator/mujoco/backends/classic_backend.py [hunk 6]
```diff
@@ -87,6 +89,50 @@ def get_render_data(self, world_id: int = 0) -> mujoco.MjData:
...
+            cam_id = mujoco.mj_name2id(self.model, mujoco.mjtObj.mjOBJ_CAMERA, name)
+            if cam_id < 0:
+                raise RuntimeError(f"Camera '{name}' not found in compiled model (expected a <camera> element).")
+            self._cam_ids[name] = cam_id
+            cam_renderers = self._mj_renderers[name] = {}
+            for data_type in cam.config.data_types:
+                renderer = mujoco.Renderer(self.model, height=cam.config.height, width=cam.config.width)
+                cam_renderers[data_type] = renderer
+                if data_type == "depth":
+                    renderer.enable_depth_rendering()
```
Comment (samuelgundry) [inline src/holosoma/holosoma/simulator/mujoco/backends/classic_backend.py:112]: defensive programming: error/warn if unsupported cnofig_type is specified (upstream could change and catch this code out)

File: src/holosoma/holosoma/simulator/mujoco/backends/classic_backend.py [hunk 7]
```diff
@@ -87,6 +89,50 @@ def get_render_data(self, world_id: int = 0) -> mujoco.MjData:
...
+
+    def render_cameras(self, cameras: list[CameraRuntime]) -> None:
+        # per-world render via the size-shared mujoco.Renderer (num_envs==1).
+        # No-hit reads the global far-clip plane (vis.map.zfar * extent, set across all cameras in
+        # the simulator's _create_sensors); remap it to +inf. The far value comes back ~0.1% under
+        # nominal (limited depth precision), so threshold at 0.99*far.
```
Comment (samuelgundry) [inline src/holosoma/holosoma/simulator/mujoco/backends/classic_backend.py:118]: tacit "why" is how you found 0.99 rather than e.g 0.9999999  
what's the risk here?  we may want to make this more user-facing later, different consumers will likely have different expectations on depth precision needs. Or perhaps even a way to log how prevalent this is for env/task X

or even opt-in mode (e.g, if someone really really needs simulator-specific behavior)

File: src/holosoma/holosoma/simulator/mujoco/backends/warp_backend.py [hunk 8]
```diff
@@ -367,6 +370,118 @@ def get_render_data(self, world_id: int = 0) -> mujoco.MjData:
...
+            for cam in cameras:
+                cam_id = self._cam_ids[cam.name]
+                if "rgb" in cam.config.data_types:
+                    rgb_mask[cam_id] = True
+                if "depth" in cam.config.data_types:
+                    depth_mask[cam_id] = True
+            self._render_context.render_rgb.assign(rgb_mask)
+            self._render_context.render_depth.assign(depth_mask)
+
+            mjw.refit_bvh(self.mjw_model, self.mjw_data, self._render_context)
```
Comment (samuelgundry) [inline src/holosoma/holosoma/simulator/mujoco/backends/warp_backend.py:472]: might need a comment here. what's this doing?  looks like a gotcha. mjwarp isn't updating its internal BVH, inverts control I guess here to let users update

File: src/holosoma/holosoma/simulator/mujoco/mujoco.py [hunk 9]
```diff
@@ -1545,6 +1554,50 @@ def find_rigid_body_indice(self, body_name: str) -> int:
...
+
+        # MuJoCo clipping is a global near/far (model.vis.map.{znear,zfar}, scaled by the scene
+        # extent), not per-camera. Honor the agnostic-core near/far across all sensor cameras by
+        # widening the global range to cover them (min near / max far). This is the same global
+        # knob the spectator video recorder tweaks (video_recorder.py).
+        assert self.root_model is not None  # compiled in load_assets, before _create_sensors
+        extent = float(self.root_model.stat.extent)
+        if extent > 0:
+            self.root_model.vis.map.znear = min(c.near for c in cameras.values()) / extent
+            self.root_model.vis.map.zfar = max(c.far for c in cameras.values()) / extent
```
Comment (samuelgundry) [inline src/holosoma/holosoma/simulator/mujoco/mujoco.py:1590]: seems problematic (?) that mjwarp has a global far plane only. thinking of eg 1024 envs for table-top manipulation (close-ups)

edit: sync'd offline. this is a global config value per-camera and not setting it to fit the global scene/env. all good.

File: src/holosoma/holosoma/simulator/plugins/ros2/worker.py [hunk 10]
```diff
@@ -0,0 +1,87 @@
+"""ROS-free drop-oldest publish worker for async sensor egress.
```
Comment (samuelgundry) [inline src/holosoma/holosoma/simulator/plugins/ros2/worker.py:1]: why do we need our own queue here? does ROS2 not have async push/pub?  could lead to surprises (or not), jsut with dropped frames, configuration, timing, etc.

**Review by samuelgundry — DISMISSED** [review]
Big effort. Cameras in Holosoma! Architecture makes sense in the context of your recent hook/plugin system _and_ compromising the existing implementations. Keen to see this in use.

**Review by samuelgundry — APPROVED** [review]
Approved, given the waiting tests are for the NEW code. Please monitor the results and we can submit a fix afterwards.
